scenario/state.py

```diff
@@ -120,6 +122,26 @@ class MetadataNotFoundError(RuntimeError):
     """Raised when Scenario can't find a metadata.yaml file in the provided charm root."""
 
 
+# This can be replaced with the KW_ONLY dataclasses functionality in Python 3.10+.
+class _MaxPositionalArgs:
+    """Raises TypeError when instantiating objects if arguments are not passed as keywords.
+
+    Looks for a `_max_positional_args` class attribute, which should be an int
+    indicating the maximum number of positional arguments that can be passed to
+    `__init__` (excluding `self`). If not present, no limit is applied.
+    """
+
+    _max_positional_args = 0
+
+    def __new__(cls, *args, **_):
+        if len(args) > getattr(cls, "_max_positional_args", float("inf")):
+            raise TypeError(
```

Comment:
Wondering if we can/should be more helpful here. We can make a pretty good guess of what attributes should be passed as keyword-only using `inspect`.

Not sure this is the best way, but:
```python
    def __new__(cls, *args, **_):
        if (n_posargs:=len(args)) > (max_posargs := getattr(cls, "_max_positional_args", float("inf"))):
            kw_only = [{name: getattr(_ann, "__name__", str(_ann))} for name, _ann in 
                       cls.__annotations__.items() if _ann is not Final][max_posargs: n_posargs]
            raise TypeError(
                f"Error instantiating {cls.__name__}. "
                f"The following arguments are keyword-only: {kw_only}",
            )
        return super().__new__(cls)
```

This gets you: 

![image](https://github.com/canonical/ops-scenario/assets/6230162/9177ac3d-e094-4e43-97c6-9cb2aa460406)

Which is a bit more helpful imho than a generic typeerror?

WDYT?

Reply:
I was just replicating what you get in Python if you use regular kwonly args (either just in a normal function or with the dataclasses.KW_ONLY thing; I assume it's the same under the hood).

But yes, we could improve on what the language does here. I'll work on that.